MySQL.Data/src/CharSetMap.cs

```diff
@@ -181,27 +180,50 @@ internal static async Task InitCollectionsAsync(MySqlConnection connection, bool
 
     internal static async Task<string> GetDefaultCollationAsync(string charset, MySqlConnection connection, bool execAsync, CancellationToken cancellationToken = default)
     {
-      SemaphoreSlim semaphoreSlim = new SemaphoreSlim(1);
-      semaphoreSlim.Wait();
-
-      if (_defaultCollations == null)
-        await InitCollectionsAsync(connection, execAsync, cancellationToken).ConfigureAwait(false);
+      if (execAsync)
+      {
+        await _semaphoreSlim.WaitAsync(cancellationToken);

```

Comment:
Missing `.ConfigureAwait(false)` introduces risk of deadlock. It should be added to all `WaitAsync` calls in this PR.

Reply:
Nice spot 👍